extensions/amp-analytics/0.1/requests.js

```diff
@@ -0,0 +1,274 @@
+
+    /** @const {!Window} */
+    this.win = ampdoc.win;
+
+    /** @const {string} */
+    this.baseUrl = dev().assert(request['baseUrl']);
+
+    /** @private {number} */
+    this.maxDelay_ = Number(request['maxDelay']) || 0;
+
+    /** @private {!./variables.VariableService} */
+    this.variableService_ = variableServiceFor(this.win);
+
+    /** @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacementService_ = Services.urlReplacementsForDoc(ampdoc);
+
+    /** @private {?Promise<string>} */
+    this.baseUrlPromise_ = null;
+
+    /** @private {?Promise<string>} */
+    this.baseUrlTemplatePromise_ = null;
+
+    /** @private {!Array<!Promise<string>>}*/
+    this.extraUrlParamsPromise_ = [];
+
+    /** @private {!../../../src/preconnect.Preconnect} */
+    this.preconnect_ = preconnect;
+
+    /** @private {!function(string, !JsonObject)} */
+    this.handler_ = handler;
+
+    /** @const @private {!Object|undefined} */
+    this.whiteList_ = isSandbox ? SANDBOX_AVAILABLE_VARS : undefined;
+
+    /** @private {?number} */
+    this.schedule_ = null;
+
+    /** @private {?JsonObject} */
+    this.lastTrigger_ = null;
+
+    /** @private {?Promise<string>} */
+    this.sendPromise_ = null;
+  }
+
+  /**
+   * Exposed method to send a request on event.
+   * Real ping may be batched and send out later.
+   * @param {?JsonObject} configParams
+   * @param {!JsonObject} trigger
+   * @param {!./variables.ExpansionOptions} expansionOption
+   * @return {!Promise<string>}
+   */
+  send(configParams, trigger, expansionOption) {
+    this.lastTrigger_ = trigger;
+    const triggerParams = trigger['extraUrlParams'];
+    if (!this.baseUrlPromise_) {
+      expansionOption.freezeVar('extraUrlParams');
+      this.baseUrlTemplatePromise_ =
+          this.variableService_.expandTemplate(this.baseUrl, expansionOption);
+      this.baseUrlPromise_ = this.baseUrlTemplatePromise_.then(baseUrl => {
+        return this.urlReplacementService_.expandAsync(
+            baseUrl, undefined, this.whiteList_);
+      });
+    };
+
+    this.extraUrlParamsPromise_.push(
+        this.expandExtraUrlParams_(configParams, triggerParams, expansionOption)
+            .then(expandExtraUrlParams => {
+              return this.urlReplacementService_.expandAsync(
+                  expandExtraUrlParams, undefined, this.whiteList_);
+            })
+            .then(finalExtraUrlParams => {
+              return finalExtraUrlParams;
+            }));
+
+    return this.trigger_();
+  }
+
+  /**
+   * Dispose function that clear reqeust handler state.
+   */
+  dispose() {
+    if (this.schedule_) {
+      this.win.clearTimeout(this.schedule_);
+    }
+    this.reset_();
+  }
+
+  /**
+   * Function that schedule the actual request send.
+   * @private
+   */
+  trigger_() {
+    if (this.maxDelay_ == 0) {
+      return this.fire_();
+    }
+    // If is batched
+    if (!this.schedule_) {
```

Comment:
nit: rename variable to timeoutId or something more descriptive.

Reply:
changed to `timetoutId_`